README.md
Connect Ra-01/Ra-02 module base on LoRaTM SX127x chip to Orange Pi
==================================================================

### Notes
0. THIS CODE IS DON'T WORK YET! WORK IN PROGRESS
1. This is experimental example, not software product ready for use
2. This this free and open source software
3. Author: Alex Zorg azorg(at)mail.ru
4. Licenced by GPLv3
5. Tested on Orange Pi Zero
6. Sources based on:
 * https://github.com/azorg/sx127x_esp
 * https://github.com/azorg/spi
 * https://github.com/azorg/sgpio
 * https://github.com/azorg/vsrpc

## Orange Pi Zero 26 pin connector

 | GPIO | Signal |Pin |Pin | Signal  | GPIO |
 |:----:| ------:|:--:|:--:|:------- |:----:|
 |      |   3.3V |  1 | 2  | 5V      |      |
 |  12  |  SDA.0 |  3 | 4  | 5V      |      |
 |  11  |  SCL.0 |  5 | 6  | 0V      |      |
 |   6  | GPIO.7 |  7 | 8  | TxD1    | 198  |
 |      |     0V |  9 | 10 | RxD1    | 199  | 
 |   1  |   RxD2 | 11 | 12 | GPIO.1  | 7    |
 |   0  |   TxD2 | 13 | 14 | 0V      |      |
 |   3  |   CTS2 | 15 | 16 | GPIO.4  | 19   |
 |      |   3.3v | 17 | 18 | GPIO.5  | 18   |
 |  15  |   MOSI | 19 | 20 | 0V      |      |
 |  16  |   MISO | 21 | 22 | RTS2    | 2    |
 |  14  |   SCLK | 23 | 24 | CE0     | 13   |
 |      |     0V | 25 | 26 | GPIO.11 | 10   |

## Connect LoRa SX1278 module to Orange Pi Zero

| GPIO | Pi Zero | Signal  | SX1278 (color)  |
|:----:|:-------:|:------- |:--------------- |
|   6  |    7    | IRQ     | DIO0  (yellow)  |
|   7  |   12    | RESET   | RESET (magenta) |
|  15  |   19    | MOSI    | MOSI  (green)   |
|  16  |   21    | MISO    | MISO  (blue)    |
|  14  |   23    | SCK     | SCK   (white)   |
|  13  |   24    | CS      | NSS   (grey)    |
|  19  |   16    | DATA *  | DIO2  (brown)   |
|  -   |   -     | DCLK ** | DIO1  (orange)  |
|  18  |   18    | LED     | -               |
|  -   | 1 or 17 | 3.3V    | 3.3V  (red)     |
|  -   | 25,20,  | GND     | GND   (black)   |
|  -   | 14,9,6  | GND     | -               |

Notes:

- (*) DIO2(DATA) is optional and may used in continuous FSK/OOK mode

- (**) DIO1(DCLK) unused in this project

- set GPIO_* and SPI_DEVICE define's in "sx127x_test.c" for your hardware

## Build test application

* edit "sx127x_test.c" module (select modes)

* build:

```
$ make
```

* run:
```
$ ./sx127x_test
```



ChangeLog
2018.10.03: Alex Zorg <azorg(at)mail.ru>
 * fix error in "sx127x" modude near packet SNR/RSSI registors

2018.09.23: Alex Zorg <azorg(at)mail.ru>
 * update sx127x.c/sx127x.h module

2018.03.28: Alex Zorg <azorg(at)mail.ru>
 * write/read 3 bytes by SPI in set_frequency()/get_frequency() at once
 - remove sx127x_spi_transfer() help funcion
 + add radio.h/radio.c layer for SBC

2018.03.27: Alex Zorg <azorg(at)mail.ru>
 + add sx127x_dump()
 * fix bug in sx127x_set_fixed()
 * update sx127x_send() prototype (add fixed)
 * LoRa/FSK/OOK tested OK

2018.03.26: Alex Zorg <azorg(at)mail.ru>
 * update sx127x_set_bitrate(): calculate fractional part
 * fix error in MODE_OOK constant
 * fix RX_BW_TABLE
 * update sx127x_receive()
 * LoRa first test OK! (explicit header mode)
 * morse demo tested
 * FSK/OOK packet mode not work yet :(

2018.03.25: Alex Zorg <azorg(at)mail.ru>
 * fix megabug in sx127x_spi_transfer()
 * fix sx127x_set_ocp()
 * first transmitter test
 * fix sx127x_set_fdev()
 * update sx127x_set_frequency()
 + add sx127x_update_band()
 + add some debug output

2018.03.23: Alex Zorg <azorg(at)mail.ru>
 * work in process
 * prepare all `sx127x_t` prototypes
 * `sx127x_t` ready to debug
 + add sx127x/README.md

2018.03.20: Alex Zorg <azorg(at)mail.ru>
 + start project



radio.h
 /*
 * -*- coding: UTF8 -*-
 * Semtech SX127x famaly chips example layer for SBC (Single Board Computer)
 * Tested on "Orange Pi Zero"
 * File: "radio.h"
 */

#ifndef RADIO_H
#define RADIO_H
//-----------------------------------------------------------------------------
#include "sx127x.h" // `sx127x_t`
//-----------------------------------------------------------------------------
#define ORANGE_PI_ZERO
//#define ORANGE_PI_ONE
//#define ORANGE_PI_WIN_PLUS

// GPIO lines and SPI devic to SX127x module
#if defined(ORANGE_PI_ZERO)
#  define RADIO_SPI_DEVICE  "/dev/spidev1.0" // on linux ver. 3.4.113-sun8i
#  define RADIO_GPIO_IRQ     6  // pin  7 of 26 (GPIO.7)
//#  define RADIO_GPIO_IRQ   1  // pin 11 of 26 (RxD2)
#  define RADIO_GPIO_RESET   7  // pin 12 of 26 (GPIO.1)
#  define RADIO_GPIO_CS      13 // pin 24 of 26 (CE0)
#  define RADIO_GPIO_DATA    19 // pin 16 of 26 (GPIO.4)
#  define RADIO_GPIO_LED     18 // pin 18 of 26 (GPIO.5)
//#  define RADIO_GPIO_LED   3  // pin 15 of 26 (CTS2)
//#  define RADIO_GPIO_LED   2  // pin 22 of 26 (RTS2)
#elif defined(ORANGE_PI_ONE)
#  define RADIO_SPI_DEVICE "/dev/spidev1.0" // FIXME
#  define RADIO_GPIO_IRQ   1 // FIXME
#  define RADIO_GPIO_RESET 2 // FIXME
#  define RADIO_GPIO_CS    3 // FIXME
#  define RADIO_GPIO_DATA  4 // FIXME
#  define RADIO_GPIO_LED   5 // FIXME
#elif defined(ORANGE_PI_WIN_PLUS)
#  define RADIO_SPI_DEVICE "/dev/spidev1.0" // FIXME
#  define RADIO_GPIO_IRQ   1 // FIXME
#  define RADIO_GPIO_RESET 2 // FIXME
#  define RADIO_GPIO_CS    3 // FIXME
#  define RADIO_GPIO_DATA  4 // FIXME
#  define RADIO_GPIO_LED   5 // FIXME
#endif
//----------------------------------------------------------------------------
// SPI max speed [Hz]
#define RADIO_SPI_SPEED 20000000 // 20 MHz 
//----------------------------------------------------------------------------
extern sx127x_t radio;
extern int radio_stop;
//----------------------------------------------------------------------------
#ifdef __cplusplus
extern "C"
{
#endif // __cplusplus
//----------------------------------------------------------------------------
// on/off DATA line
void radio_data_on(bool on);
//-----------------------------------------------------------------------------
// on/off LED
void radio_led_on(bool on);
//-----------------------------------------------------------------------------
// blink LED
void radio_blink_led();
//-----------------------------------------------------------------------------
// hard reset SX127x radio module
void radio_reset();
//----------------------------------------------------------------------------
// init SX127x radio module hardware layer (before call sx127x_init())
void radio_init();
//-----------------------------------------------------------------------------
// create listen IRQ thread (after sx127x_init())
void radio_create_irq_thread();
//-----------------------------------------------------------------------------
// free SX127x radio module
void radio_free();
//-----------------------------------------------------------------------------
// SPI exchange wrapper function (return number or RX bytes)
int radio_spi_exchange(
  u8_t       *rx_buf, // RX buffer
  const u8_t *tx_buf, // TX buffer
  u8_t len,           // number of bytes
  void *context);     // optional SPI context or NULL
//----------------------------------------------------------------------------
#ifdef __cplusplus
}
#endif // __cplusplus
//----------------------------------------------------------------------------
#endif // RADIO_H

/*** end of "radio.h" file ***/


